cirq/qis/measures.py

```diff
@@ -38,26 +46,72 @@ def fidelity(state1: np.ndarray, state2: np.ndarray) -> float:
     Args:
         state1: The first state.
         state2: The second state.
+        qid_shape: The qid shape of the given states.
+        validate: Whether to check if the given states are valid quantum states.
+
+    Returns:
+        The fidelity.
+
+    Raises:
+        ValueError: The qid shape of the given states was not specified and
+            could not be inferred.
     """
-    if len(state1.shape) == 1 and len(state2.shape) == 1:
-        # Both state vectors
-        return np.abs(np.vdot(state1, state2))**2
-    elif len(state1.shape) == 1 and len(state2.shape) == 2:
-        # state1 is a state vector and state2 is a density matrix
-        return np.real(np.conjugate(state1) @ state2 @ state1)
-    elif len(state1.shape) == 2 and len(state2.shape) == 1:
-        # state1 is a density matrix and state2 is a state vector
-        return np.real(np.conjugate(state2) @ state1 @ state2)
-    elif len(state1.shape) == 2 and len(state2.shape) == 2:
-        # Both density matrices
-        state1_sqrt = _sqrt_positive_semidefinite_matrix(state1)
-        eigs = scipy.linalg.eigvalsh(state1_sqrt @ state2 @ state1_sqrt)
-        # Zero out small negative entries
-        eigs = np.maximum(eigs, np.zeros(eigs.shape, dtype=eigs.dtype))
-        trace = np.sum(np.sqrt(eigs))
+    if isinstance(state1, int) and isinstance(state2, int):
+        return float(state1 == state2)
+
+    if isinstance(state1, value.ProductState) and isinstance(
+            state2, value.ProductState):
+        if len(state1) != len(state2):
+            raise ValueError('Mismatched number of qubits in product states: '
+                             f'{len(state1)} and {len(state2)}.')
+        return np.prod([
+            np.abs(np.vdot(s1.state_vector(), s2.state_vector()))
+            for s1, s2 in zip(state1, state2)
+        ])**2
+
+    if qid_shape is None:
+        qid_shape = infer_qid_shape(state1, state2)
+
+    state1 = quantum_state(state1,
+                           qid_shape=qid_shape,
+                           dtype=state1.dtype,
+                           validate=validate)
+    state2 = quantum_state(state2,
```

Comment:
For performance, we should not be wrapping input states to simplify control flow in the computation of fidelity. Instead, we should choose the appropriate fidelity formula based on the shape of the input numpy arrays (when the inputs are numpy arrays) and have additional branch(es) for the case when some inputs are QuantumStates. This way, most of the codepaths will completely avoid our QuantumState machinery, but at the same time in the rare cases where ambiguity arises users will be able to use `density_matrix` to tell our fidelity code what to do.

(In fact, I think we should adopt this approach more generally for all code that implements functions on quantum states, e.g. trace distance, etc.)

Reply:
I've added a special case to handle when the two inputs are complex numpy arrays.